### PR TITLE
AWLS2-444: add a NAT gateway to Azure agentless deployments by default

### PR DESCRIPTION
**Pull request by wilderj**
## Summary

This PR changes the Azure agentless networking resources to include a NAT gateway by default. We've observed that the scanning instances used in our Azure deployments frequently exhaust public IP quotas, leading to scan failures. Our current recommendation is to size these quotas in proportion to the expected size of the deployment, but customers often don't have enough information to set the quotas appropriately. 

A NAT gateway solves this problem by creating only one public IP for the deployment. Individual scanning instances no longer need an external IP of their own, as all outbound data instead transits the NAT. 

This new feature is on by default, but can be disabled via the `use_nat_gateway` variable. Use of a NAT gateways increases costs (fixed resource cost of ~$30/mo, plus additional data transit costs) so smaller customers may prefer the existing architecture. The use of this new feature will be included in upcoming changes to our external documentation.

The other networking resources we deploy needed to be slightly refactored to accommodate this change. In particular, the option to use a custom vnet/subnet now requires additional configuration -- I've updated our examples to reflect this.

## How did you test this change?

I've tested this change by deploying Azure with various network configurations:
1. With a NAT gateway and no custom vnet
2. With a NAT gateway and a custom vnet/subnet/nsg
3. With no NAT gateway and no custom vnet 
4. With no NAT gateway and a custom vnet/subnet

## Issue
https://lacework.atlassian.net/browse/AWLS2-444

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -451,13 +460,66 @@ resource "azurerm_virtual_network" "agentless_orchestrate" {
...
 }
 
+resource "azurerm_public_ip" "agentless_public_ip" {
+  depends_on = [azurerm_resource_group.scanning_rg]
+  count = var.regional && var.use_nat_gateway ? 1 : 0
+
+  name                = replace("${local.prefix}-public-ip-${local.region}-${local.suffix}", " ", "-")
+  location            = local.region
+  resource_group_name = local.scanning_resource_group_name
+  allocation_method   = "Static"
```
Comment (japsingh) [inline main.tf:490]: I see some thought has gone into deciding the sku and allocation method. The current choice seems to be the right fit.

Comment (wilderj) [reply main.tf:490]: Yeah, the allocation method was just a choice on my part. The SKU actually needs to be set, otherwise it defaults to `Basic` which is incompatible with some of the other resources we deploy. (This is actually contrary to the `azurerm` docs, which indicate that `Standard` is the default. But I digress...)

File: variables.tf [hunk 2]
```diff
@@ -72,6 +78,24 @@ variable "custom_network" {
...
+    condition = length(var.custom_network_security_group) == 0 || can(
+      regex("^/subscriptions/.*/resourceGroups/.*/providers/Microsoft.Network/networkSecurityGroups/.*$", var.custom_network_security_group)
+    )
+    error_message = "Incorrectly formatted network security group input. Use a full Resource ID in this format: /subscriptions/<subscription_id>/resourceGroups/<rg_id>/providers/Microsoft.Network/networkSecurityGroups/<nsg_id>."
+  }
+  validation {
+    condition     = (length(var.custom_network_security_group) == 0  && length(var.custom_network) == 0) || (
+      length(var.custom_network_security_group) != 0 && length(var.custom_network) != 0 && var.use_nat_gateway == true) || (
+      length(var.custom_network_security_group) == 0 && length(var.custom_network) != 0 && var.use_nat_gateway == false)
+    error_message = "You must specify a custom network security group if using a custom virtual network and a NAT gateway. "
```
Comment (japsingh) [inline variables.tf:95]: shouldn't this reflect in the custom_vnet example? Currently, I don't see custom_network_security_group being specified in that example.

Comment (wilderj) [reply variables.tf:95]: I left it commented out by default, but it's there around line 60 of `examples/custom-vnet/main.tf` with an explanatory comment.

**Review by japsingh — APPROVED** [review]
pretty important work @wilderj , this must have been quite involved.

See if we need to update the custom-vnet example, do we need to specify custom nsg there.
Also, there are some test failures blocking the merge.

**Comment by wilderj** [comment]
> pretty important work @wilderj , this must have been quite involved.
> 
> See if we need to update the custom-vnet example, do we need to specify custom nsg there. Also, there are some test failures blocking the merge.

Thanks @japsingh ! 

Concerning the test failures, I understand them now. It has to do with the way I'm validating that the customer has set the custom subnet and nsg correctly. 

Prior to Terraform 1.9, when validating a variable you could not reference a different variable from within your validation `condition`. Here it is necessary -- I need to make sure that _both_ the custom nsg and subnet are set if a NAT is used. I see two possible solution to this.

1. Simply make our minimum required Terraform version 1.9, up from 1.5 currently. This is my preferred solution, but I don't know if this will inconvenience customers. 
2. Hack around this limitation. There are ways to structure our variables differently to avoid this error, but it's not pretty. It should work but I'd rather not make this custom networking setup more complicated than it already is. 

Anyways, something to discuss at tomorrow's standup.

File: examples/custom-vnet/main.tf [hunk 3]
```diff
@@ -57,4 +57,13 @@ module "lacework_azure_agentless_scanning_rg_and_vnet" {
...
   region                         = local.region
+
+  // Creating a NAT gateway with a custom vnet requires the network security group
+  // to be specified here in the custom_network_security_group variable. If you do
+  // not wish to use a NAT gateway, do not use this variable and instead set use_nat_gateway
+  // to false. 
+
+  // Only uncomment one or the other!
+  // custom_network_security_group = azurerm_network_security_group.example.id
+  // use_nat_gateway = false
```
Comment (kirklandnuts) [inline examples/custom-vnet/main.tf:68]: nit: Perhaps we should set `custom_network_security_group` by default, aligning with the idea of "convention over configuration".

I understand that the intention of leaving both `use_nat_gateway` and `custom_network_security_group` commented here is to clarify that they're mutually exclusive. 

However, since the default is `use_nat_gateway = true`, does it make sense to have `custom_network_security_group` set by default to align with that? In this case, if someone wants to disable the NAT gateway, they will also unset `custom_network_security_group` — otherwise, the validation will trigger and clarify that only one of these two variables should be set at a time. 

We could tweak the comments to remain explicit while still defaulting to the most common case.
```suggestion
  // When using a custom vnet with the default NAT gateway (use_nat_gateway = true),
  // you must specify the network security group here:
  custom_network_security_group = azurerm_network_security_group.example.id

  // If you want to use public IPs instead of a NAT gateway, comment out the line above
  // and uncomment this line:
  // use_nat_gateway = false
```

IMO, this approach provides better developer experience for the most common use case as it would work out of the box, and we would still retain safety provided by the validation rules you added.

What do you think?

**Review by kirklandnuts — APPROVED** [review]
nice Joe

I had a non-blocking comment about aligning the vnet example with the default behavior — otherwise, this LGTM 👍

File: variables.tf (hunk 2 above)
Comment (kirklandnuts) [reply variables.tf:95]: I shared my thoughts on this [here](https://github.com/lacework/terraform-azure-agentless-scanning/pull/43#discussion_r1890741369)

**Comment by kirklandnuts** [comment]
> Prior to Terraform 1.9, when validating a variable you could not reference a different variable from within your validation `condition`. Here it is necessary -- I need to make sure that _both_ the custom nsg and subnet are set if a NAT is used. I see two possible solution to this.
> 
> 1. Simply make our minimum required Terraform version 1.9, up from 1.5 currently. This is my preferred solution, but I don't know if this will inconvenience customers.
> 2. Hack around this limitation. There are ways to structure our variables differently to avoid this error, but it's not pretty. It should work but I'd rather not make this custom networking setup more complicated than it already is.
> 
> Anyways, something to discuss at tomorrow's standup.

I agree, approach #1 seems more ideal

File: examples/custom-vnet/main.tf (hunk 3 above)
Comment (wilderj) [reply examples/custom-vnet/main.tf]: I like it! 

The only reason I had both options commented out by default was because above on line 57 we also leave the `custom_network` commented out. I presumed this was because trying to run `terraform apply` will result in an error if the vnet doesn't already exist. So I was just trying to follow the precendent.

But on further consideration I don't really love that -- it's confusing and the customer has to run `terraform apply` twice regardless. 

So TL;DR I've committed your suggestion, but also uncommented `custom_network` above.
